### PR TITLE
Add support for error unwrapping

### PR DESCRIPTION
**Pull request by hwh33**
With this PR, any instances of errors.Error _which have an associated cause_ will implement the method `Unwrap() error`.  Errors with no cause will not implement this method.  This means that `errors.Is` and `errors.As` will work on the errors produced by this package in the same way they would for errors produced by `fmt.Errorf` or `errors.New`.  For concrete examples, see the new test, `TestUnwrapping`.

The public API is completely unchanged, so this can serve as a drop-in replacement.

### CODE REVIEW

**Comment by hwh33** [comment]
For justification on updating this API vs just using the standard library's error wrapping API (`fmt.Errorf("something happened: %w", err)`), see [this comment](https://github.com/getlantern/lantern-internal/issues/4127#issuecomment-696828744)

**Comment by hwh33** [comment]
The diff is a little hard to parse, so here is a little introduction: I split the concrete error type into two types: `baseError` and `wrappingError`.  The latter embeds the former, so a `wrappingError` is also a `baseError`, it just has a cause attached.

Both `New` and `Wrap` can return an error of either type, depending on whether a cause is found in the input.  Of course, callers will only see the `Error` interface, so it all looks the same to them.

I made only minor changes to the tests to ensure that API behavior is unchanged.  I believe the only changes I made were:
- switches from `%v` to `%w` in `fmt.Errorf` calls
- changes to type-checking to accommodate the now-split concrete type

File: errors.go [hunk 1]
```diff
@@ -333,8 +295,111 @@ func buildError(desc string, fullText string, wrapped error, cause Error) *struc
...
+// Implements error unwrapping as described in the standard library's errors package:
+// https://golang.org/pkg/errors/#pkg-overview
+func (e *wrappingError) Unwrap() error {
+	return e.wrapped
+}
+
+func (e *wrappingError) Fill(m context.Map) {
+	type filler interface{ Fill(context.Map) }
+
+	if f, ok := e.wrapped.(filler); ok {
```
Comment (oxtoacart) [inline errors.go:327]: What if you have one of our errors wrapped in a standard Go error? I think maybe we need to unwrap as far as we can and look for all `filler`s along the way.

Comment (hwh33) [reply errors.go]: That's a good point.  Fixed here: https://github.com/getlantern/errors/pull/6/commits/ee0d19731faa77ba4bda428fdb87d722524cd52e

Comment (hwh33) [reply errors.go]: This is cleaner I think: https://github.com/getlantern/errors/pull/6/commits/8fa0b61f584e2d32c84c076ff76da071430aac68

File: errors_test.go [hunk 2]
```diff
@@ -146,6 +146,25 @@ func TestHiddenWithCause(t *testing.T) {
 	// doesn't panic. If we get to this point without panicking, we're happy.
 }
 
+func TestFill(t *testing.T) {
+	e := New("something happened").(*baseError)
+	e2 := New("uh oh: %v", e).(*wrappingError)
```
Comment (oxtoacart) [inline errors_test.go:151]: Might be nice to include a standard Go error in the chain to make sure that keeps working.

**Review by oxtoacart — APPROVED** [review]
LGTM

File: errors_test.go (hunk 2 above)
Comment (hwh33) [reply errors_test.go:151]: That's smart.  Just added that.

**Comment by hwh33** [comment]
Thanks @oxtoacart!
